### PR TITLE
p4ctf finals 2023

### PR DESCRIPTION
**Pull request by Pharisaeus**

### CODE REVIEW

File: 2023-06-21-p4ctf-finals/rsa_kebab/README.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+### Modular root approach
+
+As mentioned before, `y` is small and we can easily brute-force through all possible values.
+This means we can multiply by `modinv(3^i, 2**512)` to "cancel out" this factor if we happen to match `y==i`.
+
+We get `n * modinv(3^y, 2**512) 2**512 = p**2 mod 2**512`
+
+Now we can simply take modular square root of this value to get back `p`.
+Normally taking modular roots over composites require taking root mod each prime and then combining results via CRT.
+Here w have a special case becase we have repeated primes in the modulus factorization and we need to use Hensel Lifting to raise solution mod `prime` to solution mod `prime^k`.
```
Comment (XeR) [inline 2023-06-21-p4ctf-finals/rsa_kebab/README.md:60]:
```suggestion
Here we have a special case becase we have repeated primes in the modulus factorization and we need to use Hensel Lifting to raise solution mod `prime` to solution mod `prime^k`.
```

File: 2023-06-21-p4ctf-finals/rsa_kebab/README.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+It's worth pointing out that `y` in this case is relatively small (in the order of hundreds).
+
+This means that modulus `n = p*q = p * (p*3^y mod 2**512)`
+
+## Solution
+
+There are 2 ways to approach this task.
+In both cases we would like to first simplify the modulus equation a bit.
+
+We notice that of we take `n % 2**512` we can simplify the modulus to:
```
Comment (XeR) [inline 2023-06-21-p4ctf-finals/rsa_kebab/README.md:47]:
```suggestion
We notice that if we take `n % 2**512` we can simplify the modulus to: 
```

File: 2023-06-21-p4ctf-finals/narco_chat/README.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+There are lots of potential avenues to look into:
+
+1. Maybe we can register the same user twice by making "equivalent" base64 using unused bits?
+2. Maybe we can register the same user twice by encrypting some `\x00` or `\n` or something similar?
+3. Maybe we can somehow predict the user IDs by figuring out how the hash is computed (eg. maybe some md5(username+secret))
+
+The actual solution we found purely by accident. 
+While sending random requests we at some point got back "DB error" message.
+This pointed us to potential SQLi when trying to read messages.
+We could do a classic `'or'1'='1` trick to read all messages, not only those beloging to our user and getting the flag: `p4{Was_this_a_crypto?a_web?a_re?who_the_hell_knows}`
```
Comment (XeR) [inline 2023-06-21-p4ctf-finals/narco_chat/README.md:42]:
```suggestion
We could do a classic `'or'1'='1` trick to read all messages, not only those belonging to our user and getting the flag: `p4{Was_this_a_crypto?a_web?a_re?who_the_hell_knows}`
```

**Review by XeR — COMMENTED** [review]
I'll push the changes on your branch (if I can). There is still one comment about a missing mod where I need your feedback (I don't speak numbers)

File: 2023-06-21-p4ctf-finals/rsa_kebab/README.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+We notice that of we take `n % 2**512` we can simplify the modulus to: 
+
+`n mod 2**512 = p * (p*3^y mod 2**512) mod 2**512 = p**2 * 3^y mod 2**512 = 3**(2x+y) mod 2**512`
+
+### Modular root approach
+
+As mentioned before, `y` is small and we can easily brute-force through all possible values.
+This means we can multiply by `modinv(3^i, 2**512)` to "cancel out" this factor if we happen to match `y==i`.
+
+We get `n * modinv(3^y, 2**512) 2**512 = p**2 mod 2**512`
```
[the remark this answers is not in the document]
Comment (Pharisaeus) [reply 2023-06-21-p4ctf-finals/rsa_kebab/README.md:56]: Yes

**Review by XeR — APPROVED** [review]

**Pull request merged**